File: 백준/Silver/14402. 야근/야근.py
from collections import defaultdict
import sys
input = sys.stdin.readline

data = defaultdict(list)
name_counter = defaultdict(int)
q = int(input())
for _ in range(q):
    s, p = input().strip().split()
    data[s].append(p)

# print(data)
ans = 0
for name, history in data.items():
    for action in history:
        if action == '+':
            name_counter[name] += 1
        else:
            if name_counter[name] == 0:
                ans += 1
            else:
                name_counter[name] -= 1
    ans += name_counter[name]
    
print(ans)

File: 백준/Silver/14402. 야근/야근.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        Map<String, List<String>> data = new HashMap<>();
        Map<String, Integer> nameCounter = new HashMap<>();
        
        int q = Integer.parseInt(br.readLine());

        for (int i = 0; i < q; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String s = st.nextToken();
            String p = st.nextToken();
            
            data.putIfAbsent(s, new ArrayList<>());
            data.get(s).add(p);
        }
        
        int ans = 0;
        
        for (String name : data.keySet()) {
            List<String> history = data.get(name);
            
            for (String action : history) {
        	if (action.equals("+")) {
        	    nameCounter.put(name, nameCounter.getOrDefault(name, 0) + 1);
        	} else {
        	    if (nameCounter.getOrDefault(name, 0) == 0) {
        		ans++;
        	    } else {
        		nameCounter.put(name, nameCounter.get(name) - 1);
        	    }
        	}
            }
            
            ans += nameCounter.getOrDefault(name, 0);
        }
        
        System.out.println(ans);
    }
}